modules/miner/update.go

```diff
@@ -28,11 +26,139 @@ func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
+			// Place candidate into block,
+			m.blockMapHeap.Push(elem)
+
+			// Place transactions removed from block heap into
+			// the overflow heap.
+			for _, v := range bottomSets {
+				m.overflowMapHeap.Push(v)
+			}
+			break
+		}
+
+		// If the blockMapHeap is empty, push all elements removed from it back
+		// in, and place the candidate set into the overflow. This should never
+		// happen since transaction sets are much smaller than the max block
+		// size.
+		_, exists := m.blockMapHeap.Peek()
+		if !exists {
+			m.overflowMapHeap.Push(elem)
+			// Put back in transactions removed.
+			for _, v := range bottomSets {
+				m.blockMapHeap.Push(v)
+			}
+			// Finished with this candidate set.
+			break
+		}
+		// Add the set to the bottomSets slice. Note that we don't increase
+		// sizeOfBottomSets until after calculating the average.
+		nextSet := m.blockMapHeap.Pop()
+		bottomSets = append(bottomSets, nextSet)
+
+		// Calculating fees to compare total fee from those sets removed and the current set s.
+		totalFeeFromNextSet := nextSet.set.averageFee.Mul64(nextSet.set.size)
+		totalBottomFees := averageFeeOfBottomSets.Mul64(sizeOfBottomSets).Add(totalFeeFromNextSet)
+		sizeOfBottomSets += nextSet.set.size
+		averageFeeOfBottomSets := totalBottomFees.Div64(sizeOfBottomSets)
+
+		// If the average fee of the bottom sets from the block is higher than
+		// the fee from this candidate set, put the candidate into the overflow
+		// MapHeap.
+		if averageFeeOfBottomSets.Cmp(candidateSet.averageFee) == 1 {
+			// CandidateSet goes into the overflow.
+			m.overflowMapHeap.Push(elem)
+			// Put transaction sets from bottom back into the blockMapHeap.
+			for _, v := range bottomSets {
+				m.blockMapHeap.Push(v)
+			}
+			// Finished with this candidate set.
+			break
+		}
+	}
+}
+
+// addNewTxns adds new unconfirmed transactions to the miner's transaction
+// selection and updates the splitSet and mapElement state of the miner.
+func (m *Miner) addNewTxns(diff *modules.TransactionPoolDiff) {
+	// Get new splitSets (in form of mapElement)
+	newElements := m.getNewSplitSets(diff)
+
+	// Place each elem in one of the MapHeaps.
+	for i := 0; i < len(newElements); i++ {
+		// Add splitSet to miner's global state using pointer and ID stored in
+		// the mapElement and then add the mapElement to the miner's global
+		// state.
+		m.splitSets[newElements[i].id] = newElements[i].set
+		m.addMapElementTxns(newElements[i])
+	}
+}
+
+// Change the UnsolvedBlock so that it  has exactly those transactions in the
+// blockMapHeap.
+func (m *Miner) adjustUnsolvedBlock() {
+	numTxns := 0
+	for _, elem := range m.blockMapHeap.selectID {
+		numTxns += len(elem.set.transactions)
+	}
+	// If the transactions that need to be added don't fit in the block,
+	// increase the size of the block by a constant factor to be more efficient.
+	if numTxns > cap(m.persist.UnsolvedBlock.Transactions) {
+		newCap := cap(m.persist.UnsolvedBlock.Transactions) * 6 / 5
```

Comment:
what is the initial size of `Transactions`?
FYI the default Go slice behavior is to double until 1024 elements, and grow by 25% thereafter.